src/server/pfs/s3/bucket.go

```diff
@@ -0,0 +1,385 @@
+package s3
+
+import (
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/gorilla/mux"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/server/pkg/errutil"
+)
+
+// this is a var instead of a const so that we can make a pointer to it
+var defaultMaxKeys int = 1000
+
+const locationSource = `
+<LocationConstraint xmlns="http://s3.amazonaws.com/doc/2006-03-01/">PACHYDERM</LocationConstraint>`
+
+const listObjectsSource = `
+<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+	<Name>{{ .bucket }}</Name>
+	<Prefix>{{ .prefix }}</Prefix>
+	<Marker>{{ .marker }}</Marker>
+	<MaxKeys>{{ .maxKeys }}</MaxKeys>
+	<IsTruncated>{{ .isTruncated }}</IsTruncated>
+	{{ range .files }}
+		<Contents>
+			<Key>{{ .File.Path }}</Key>
+			<LastModified>{{ formatTime .Committed }}</LastModified>
+			<ETag></ETag>
+			<Size>{{ .SizeBytes }}</Size>
+			<StorageClass>STANDARD</StorageClass>
+			<Owner>
+				<ID>000000000000000000000000000000</ID>
+				<DisplayName>pachyderm</DisplayName>
+			</Owner>
+		</Contents>
+	{{ end }}
+	{{ if .dirs }}
+		{{ range .dirs }}
+			<CommonPrefixes>
+				<Prefix>{{ . }}/</Prefix>
+			</CommonPrefixes>
+		{{ end }}
+	{{ end }}
+</ListBucketResult>`
```

Comment:
This might be a little bit cleaner if you used the `encoding/xml` package. I think you could just define a struct that you fill in, similar to the `map[string]interface{}` below and it'd serialize the right thing for you. I'm not sure if you need more control over the format than that gives you though.

Reply:
I think that gives me all the control I need, but I was concerned about boilerplate, since there's a lot of hardcoded values I put in to match S3's XML output. I'll experiment with `encoding/xml`, though, once multipart is wrapped up.